packages/@aws-cdk/aws-codebuild/lib/project.ts

```diff
@@ -1552,37 +1553,134 @@ export class WindowsBuildImage implements IBuildImage {
   }
 
   public validate(buildEnvironment: BuildEnvironment): string[] {
-    const ret: string[] = [];
-    if (buildEnvironment.computeType === ComputeType.SMALL) {
-      ret.push('Windows images do not support the Small ComputeType');
-    }
-    return ret;
+    return validateWindowsBuildEnvironment(buildEnvironment);
   }
 
   public runScriptBuildspec(entrypoint: string): BuildSpec {
-    return BuildSpec.fromObject({
-      version: '0.2',
-      phases: {
-        pre_build: {
-          // Would love to do downloading here and executing in the next step,
-          // but I don't know how to propagate the value of $TEMPDIR.
-          //
-          // Punting for someone who knows PowerShell well enough.
-          commands: [],
-        },
-        build: {
-          commands: [
-            'Set-Variable -Name TEMPDIR -Value (New-TemporaryFile).DirectoryName',
-            `aws s3 cp s3://$env:${S3_BUCKET_ENV}/$env:${S3_KEY_ENV} $TEMPDIR\\scripts.zip`,
-            'New-Item -ItemType Directory -Path $TEMPDIR\\scriptdir',
-            'Expand-Archive -Path $TEMPDIR/scripts.zip -DestinationPath $TEMPDIR\\scriptdir',
-            '$env:SCRIPT_DIR = "$TEMPDIR\\scriptdir"',
-            `& $TEMPDIR\\scriptdir\\${entrypoint}`,
-          ],
-        },
-      },
+    return runScritpWindowsBuildSpec(entrypoint);
+  }
+}
+
+/**
+ * A CodeBuild image running Windows Server 2019.
+ *
+ * This class has a bunch of public constants that represent the most popular images.
+ *
+ * You can also specify a custom image using one of the static methods:
+ *
+ * - Windows2019BuildImage.fromDockerRegistry(image[, { secretsManagerCredentials }])
+ * - Windows2019BuildImage.fromEcrRepository(repo[, tag])
+ * - Windows2019BuildImage.fromAsset(parent, id, props)
+ *
+ * @see https://docs.aws.amazon.com/codebuild/latest/userguide/build-env-ref-available.html
+ */
+export class Windows2019BuildImage implements IBuildImage {
```

Comment:
While there's nothing wrong with this solution, I have a different idea that will reduce the duplication between `WindowsBuildImage` and `Windows2019BuildImage`.

1. Let's remove `Windows2019BuildImage` completely for now. Everything will be part of `WindowsBuildImage`.
2. Let's create a public, exported enum, `WindowsImageType`. It will have 2 members: `STANDARD`, with value `'WINDOWS_CONTAINER'`, and `SERVER_2019`, with value `'WINDOWS_SERVER_2019_CONTAINER'`.
3. Add an optional property `imageType?: WindowsImageType` to `WindowsBuildImageProps`.
4. In `WindowsBuildImage`, change `type` from being hard-coded in the property declaration, to being assigned in the constructor: `this.type = props.imageType?.toString() ?? WindowsImageType.STANDARD.toString();`.
5. Have `WIN_SERVER_CORE_2019_BASE` be a constant in `WindowsBuildImage`:

```ts
  public static readonly WIN_SERVER_CORE_2019_BASE: IBuildImage = new WindowsBuildImage({
    imageId: 'aws/codebuild/windows-base:2019-1.0',
    imagePullPrincipalType: ImagePullPrincipalType.CODEBUILD,
    imageType: WindowsImageType.SERVER_2019,
  });
```

6. Change `WindowsBuildImage.fromDockerRegistry()` to take a new interface, `WindowsDockerImageOptions`, instead of `DockerImageOptions`:

```ts
export interface WindowsDockerImageOptions extends DockerImageOptions {
  /** @default WindowsImageType.STANDARD */
  readonly imageType?: WindowsImageType;
}
```

7. In the implementation of `WindowsBuildImage.fromDockerRegistry()`, pass `options.imageType` when creating a new `WindowsBuildImage`.

Reply:
I will proceed to implement this change. Regarding the way we fix `fromDockerRegistry` with the `WindowsDockerImageOptions`, we also have `fromEcrRepository` and `fromAsset`... should I also add an `imageType?` property to a subclass of `DockerImageAssetProps` for the latter? For the ECR case, I don't see a way to get this information from the `ecr.IRepository`.

Reply:
Yes, all 3 should be changed in a similar fashion, but I wouldn't do it by adding a new optional argument everywhere 🙂.

Same like I suggested creating `interface WindowsDockerImageOptions extends DockerImageOptions` for `fromDockerRegistry()`, I suggest creating `interface WindowsDockerImageAssetProps extends DockerImageAssetProps` for `fromAsset()`.

The `fromEcrRepository()` is a harder one to crack, because of the default parameter already there... I think for that one, I'm fine with adding a third parameter with a default value, like you suggested.

Reply:
One difference here is that by using the `Wndows*` classes we would change the signature of the method, and this would be a breaking/non backwards compatible change (a CDK user that had written code using `DockerImageOptions` would see its code break after update since the signature would have changed). By contrast, looking at the Java version of the `fromEcrRepository()` function, which has a default parameter `tag`, I see that if we introduce a default parameter, the method is overloaded, so there are [two versions of the function](https://docs.aws.amazon.com/cdk/api/latest/java/software/amazon/awscdk/services/codebuild/WindowsBuildImage.html#fromEcrRepository-software.amazon.awscdk.services.ecr.IRepository-), one with and one without the default parameter.